usb: device_next: support BOS descriptor with vendor request code

jfischer-no opened the pull request:
Support BOS descriptor with vendor request code and add WebUSB sample for the new USB device stack.

jfischer-no committed:
usb: device_next: support vendor request with recipient device

Allow the user to register a vendor request node identified by the
vendor code (bRequest) and containing two callbacks to handle the vendor
request. The device stack uses the vendor request node to call the
vendor request callbacks when it receives a request of type Vendor,
recipient Device, and bRequest value equal to the vendor code.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: usb: add samples function to setup USB device only

Add a function similar to sample_usbd_init_device(), but one that does
not initialize the device. It allows the application to set additional
features, such as additional descriptors.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: usb: add new WebUSB sample

Add a WebUSB sample that uses the new USB device support.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic commented on subsys/usb/device_next/usbd_ch9.c:
```diff
@@ -913,6 +913,55 @@ static int std_request_to_host(struct usbd_context *const uds_ctx,
+{
+	struct usbd_desc_node *desc_nd;
+
+	SYS_DLIST_FOR_EACH_CONTAINER(&uds_ctx->descriptors, desc_nd, node) {
```
Why abuse the descriptors list? I think it would be better to just have a separate list for the vendor requests. The vendor requests should not be limited to BOS-only, but generally any vendor (here: application developer) should be able to provide their own vendor commands in application.

jfischer-no replied:
> Why abuse the descriptors list? 

It is a descriptor and it does not abusing anything.

> I think it would be better to just have a separate list for the vendor requests. The vendor requests should not be limited to BOS-only, but generally any vendor (here: application developer) should be able to provide their own vendor commands in application.

There are no requirements in the specification to allow arbitrary vendor request with recipient device. Using the BOS platform capability is the most portable.

tmon-nordic replied:
How is this a descriptor? This is not descriptor in any way, just a way for application to register vendor commands. USB specification doesn't require the devices to implement any optional functionality (vendor commands are optional functionality). In this context the specification states "A device vendor may also define requests supported by the device.".

What do you mean by "most portable"? What does portability have to do with a stack functionality to allow application developer to support vendor commands?

jfischer-no replied:
Please familiarize yourself with how MSOS2 and WebUSB use the BOS descriptor, I am pretty sure you will understand it.

tmon-nordic commented on samples/subsys/usb/webusb/src/msosv2.h:
```diff
@@ -0,0 +1,116 @@
+		/* Microsoft OS 2.0 descriptor platform capability UUID
+		 * D8DD60DF-4589-4CC7-9CD2-659D9E648A9F
+		 */
+		.PlatformCapabilityUUID = {
+			0xDF, 0x60, 0xDD, 0xD8,
+			0x89, 0x45,
+			0xC7, 0x4C,
+			0x9C, 0xD2,
+			0x65, 0x9D, 0x9E, 0x64, 0x8A, 0x9F,
+		},
+	},
+	.cap = {
+		.dwWindowsVersion = sys_cpu_to_le32(SAMPLE_MSOS2_OS_VERSION),
+		.wMSOSDescriptorSetTotalLength = sys_cpu_to_le16(sizeof(msosv2_desc)),
+		.bMS_VendorCode = SAMPLE_MSOS2_VENDOR_CODE,
+		.bAltEnumCode = 0x00
+	},
+};
+
+static int msosv2_to_host_cb(const struct usbd_context *const ctx,
+			     const struct usb_setup_packet *const setup,
+			     struct net_buf *const buf)
+{
+	LOG_INF("Vendor callback to host");
+
+	if (setup->bRequest == SAMPLE_MSOS2_VENDOR_CODE &&
+	    setup->wIndex == MS_OS_20_DESCRIPTOR_INDEX) {
+		LOG_INF("Get MS OS 2.0 Descriptor Set");
+
+		net_buf_add_mem(buf, &msosv2_desc,
+				MIN(net_buf_tailroom(buf), sizeof(msosv2_desc)));
+
+		return 0;
+	}
+
+	return -ENOTSUP;
+}
+
+USBD_DESC_BOS_VREQ_DEFINE(bos_vreq_msosv2, sizeof(bos_msosv2_desc), &bos_msosv2_desc,
+			  SAMPLE_MSOS2_VENDOR_CODE, msosv2_to_host_cb, NULL);
```
Why this is defined in a header?

jfischer-no replied:
Why not? Or what is the problem? This is limited to a specific part and can be used as a template.

tmon-nordic replied:
Everything compilation unit that includes this header will have the variables in the corresponding object file.

jfischer-no replied:
There is only that include it. So where is the problem?

tmon-nordic commented on samples/subsys/usb/webusb/src/webusb.h:
```diff
@@ -0,0 +1,90 @@
+}
+
+USBD_DESC_BOS_VREQ_DEFINE(bos_vreq_webusb, sizeof(bos_cap_webusb), &bos_cap_webusb,
+			  SAMPLE_WEBUSB_VENDOR_CODE, webusb_to_host_cb, NULL);
```
Why this is defined in a header?

tmon-nordic commented on samples/subsys/usb/webusb/src/msosv2.h:
```diff
@@ -0,0 +1,116 @@
+	'3', 0x00, '9', 0x00, '5', 0x00, '6', 0x00, '2', 0x00, 'B', 0x00, \
+	'3', 0x00, '}', 0x00, 0x00, 0x00, 0x00, 0x00
+
+struct msosv2_descriptor {
```
Would it be possible to somehow take the advantage of the new stack to enable automatic `struct msosv2_function_subset_header` insertion when there is more than 1 interface?

The old stack did have https://github.com/zephyrproject-rtos/zephyr/pull/66449 as a solution, but that's fragile approach.

jfischer-no replied:
I do not think we can handle it. IMO it is better to write a comprehensive documentation about it, with some examples. I think the average user has no idea where and how to use MSOSv2 descriptors.

tmon-nordic replied:
Comprehensive documentation is not a real solution but a way to avoid having to come up with a real solution. While the average user might not have idea about individual MSOSv2 descriptors, it would be good to provide "just-working" experience (if possible).

henrikbrixandersen reviewed (commented):
I would prefer a generic way to register vendor requests targeting the device (not a class nor an interface) from the application code. This could easily be used for handling MSOSv2 vendor requests, but wouldn't be limited to that (not all vendor requests would require a BOS).

jfischer-no committed:
usb: device_next: support BOS descriptor with vendor request code

Platform capability descriptors such as MSOSv2 or WebUSB BOS have a
vendor request code that is used by the host to perform vendor-specific
requests. Add a convenient way to define and register a platform
capability descriptor with a vendor request node.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

carlescufi commented:
@tmon-nordic @henrikbrixandersen can you please take another look?

tmon-nordic commented on samples/subsys/usb/webusb-next/src/msosv2.h:
```diff
@@ -0,0 +1,116 @@
+	'3', 0x00, '}', 0x00, 0x00, 0x00, 0x00, 0x00
+
+struct msosv2_descriptor {
+	struct msosv2_descriptor_set_header header;
```
If not implement it right now, at least add a TODO: Add subset header when there is more than one interface.

See #66449 for more information.

jfischer-no replied:
I have aligned it with the descriptor in the legacy sample and added a comment.

tmon-nordic reviewed (changes requested):
The new bRequest registration seems fine, just few things to fix.

jfischer-no commented on include/zephyr/usb/usbd.h:
```diff
@@ -597,6 +647,21 @@ static inline void *usbd_class_get_private(const struct usbd_class_data *const c
+ * @param vto_dev   Vendor callback for to-device direction request
+ */
+#define USBD_VREQUEST_DEFINE(name, vcode, vto_host, vto_dev)			\
+	static struct usbd_vreq_node name = {					\
```
Then the user does not know the real structure name. This makes sense for more abstract macros, but not here.

carlescufi commented:
@henrikbrixandersen PTAL
